fem/bilininteg_convection_pa.cpp

```diff
@@ -13,110 +13,109 @@
-   const int NQ = Q1D*Q1D;
-   auto W = w.Read();
+   constexpr int DIM = 2;
 
-   auto J = Reshape(j.Read(), NQ, 2, 2, NE);
-   const bool const_v = vel.Size() == 2;
-   auto V =
-      const_v ? Reshape(vel.Read(), 2,1,1) : Reshape(vel.Read(), 2,NQ,NE);
-   auto y = Reshape(op.Write(), NQ, 2, NE);
+   const bool const_v = vel.Size() == DIM;
 
-   MFEM_FORALL(e, NE,
+   const auto W = w.Read();
+   const auto J = Reshape(j.Read(), NQ,DIM,DIM,NE);
+   const auto V = const_v ?
+                  Reshape(vel.Read(), DIM,1,1) :
+                  Reshape(vel.Read(), DIM,NQ,NE);
+   auto y = Reshape(op.Write(), NQ,DIM,NE);
+
+   MFEM_FORALL(q_global, NE*NQ,
    {
-      for (int q = 0; q < NQ; ++q)
-      {
-         const double J11 = J(q,0,0,e);
-         const double J21 = J(q,1,0,e);
-         const double J12 = J(q,0,1,e);
-         const double J22 = J(q,1,1,e);
-         const double w = alpha * W[q];
-         const double v0 = const_v ? V(0,0,0) : V(0,q,e);
-         const double v1 = const_v ? V(1,0,0) : V(1,q,e);
-         const double wx = w * v0;
-         const double wy = w * v1;
-         //w*J^-1
-         y(q,0,e) =  wx * J22 - wy * J12; // 1
-         y(q,1,e) = -wx * J21 + wy * J11; // 2
-      }
+      const int e = q_global / NQ;
+      const int q = q_global % NQ;
+      const double J11 = J(q,0,0,e);
+      const double J21 = J(q,1,0,e);
+      const double J12 = J(q,0,1,e);
+      const double J22 = J(q,1,1,e);
+      const double w = alpha * W[q];
+      const double v0 = const_v ? V(0,0,0) : V(0,q,e);
+      const double v1 = const_v ? V(1,0,0) : V(1,q,e);
+      const double wx = w * v0;
+      const double wy = w * v1;
+      // y = alpha * W * det(J) * J^{-1} . v = adj(J) . { wx, wy }
+      y(q,0,e) =  wx * J22 - wy * J12; // 1
+      y(q,1,e) = -wx * J21 + wy * J11; // 2
    });
 }
 
 // PA Convection Assemble 3D kernel
-static void PAConvectionSetup3D(const int Q1D,
+static void PAConvectionSetup3D(const int NQ,
                                 const int NE,
                                 const Array<double> &w,
                                 const Vector &j,
                                 const Vector &vel,
                                 const double alpha,
                                 Vector &op)
 {
-   const int NQ = Q1D*Q1D*Q1D;
-   auto W = w.Read();
-   auto J = Reshape(j.Read(), NQ, 3, 3, NE);
-   const bool const_v = vel.Size() == 3;
-   auto V =
-      const_v ? Reshape(vel.Read(), 3,1,1) : Reshape(vel.Read(), 3,NQ,NE);
-   auto y = Reshape(op.Write(), NQ, 3, NE);
-   MFEM_FORALL(e, NE,
+   constexpr int DIM = 3;
+   constexpr int SDIM = DIM;
+   const auto W = Reshape(w.Read(), NQ);
+   const auto J = Reshape(j.Read(), NQ,SDIM,DIM,NE);
+   const bool const_v = vel.Size() == DIM;
+   const auto V = const_v ?
+                  Reshape(vel.Read(), 3,1,1) :
+                  Reshape(vel.Read(), 3,NQ,NE);
+   auto y = Reshape(op.Write(), NQ,3,NE);
+   MFEM_FORALL(q_global, NE*NQ,
```

Comment:
Maybe thread the inner loops?

Reply:
Which inner loop? Since each quadrature point is independent this should be fast.

Reply:
Oh I see it now, 1D index gets converted to 2D here: 
```
const int e = q_global / NQ;
const int q = q_global % NQ;
```